sdk/Pulumi.Automation/Events/GrpcEventWatcher.cs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2016-2025, Pulumi Corporation
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Hosting.Server;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Pulumi.Automation.Events
+{
+    /// <summary>
+    /// gRPC-based event watcher that hosts a gRPC server to receive engine events.
+    /// </summary>
+    internal sealed class GrpcEventWatcher : IEventWatcher
+    {
+        private readonly IHost _host;
+        private readonly TaskCompletionSource<int> _portTcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly CancellationTokenSource _shutdownCts = new CancellationTokenSource();
+
+        public string EventLogPath { get; private set; } = string.Empty;
+
+        public GrpcEventWatcher(Action<EngineEvent> onEvent, CancellationToken cancellationToken)
+        {
+            _host = Host.CreateDefaultBuilder()
```

Comment:
We've got a few places now that repeat this server setup code, feels like we should probably put this in a util class now. I can do that later if you don't want to muck about with C#.

Reply:
Yeah, makes sense.  I'd rather not do it in this PR, but opened https://github.com/pulumi/pulumi-dotnet/issues/738.